sourcegraph/sourcegraph
# Code Insights: Improve dashboard select UI

## Problem

This PR improves dashboard select UI on the code insight dashboard page. Previously this select was built with [Listbox UI from reach-ui](https://reach.tech/listbox). And it was good enough but we had a few problems with 
- Keyboard navigation since listbox doesn't work well with anything that is not listbox component (it's bad because we want to have search filter input within listbox popover)
- Listbox options accessibility. Keyboard navigation doesn't support scroll into view functionality for focused/active listbox options.
- Listbox Popover has to be fixed size and doesn't have smart position logic that we have with out in-house popover component
- Filter highlighting logic, we want to highlight parts that match with filter value in the input and we had to write this logic ourselves back then. 

## Solution

In this PR we replace Listbox UI with combination of wildcard Popover and Combobox UI. This gives us better UI/UX. 

<img width="1064" alt="Screenshot 2022-10-17 at 18 11 14" src="https://user-images.githubusercontent.com/18492575/196284192-d6879fd5-49dd-4e09-9280-a160c697cad3.png">

- Better keyboard navigation. All options are accessible from keyboard. It works nicely with search filter input (it has autocompletion as you navigate through options) 
- Better Popover logic. We use our in-house popover this means that select UI popover will fit in the visible viewport area based on content and available area below the select element.
- Slightly updated UI. Dashboard select UI is built now with more or less the same layout concept as all other comboboxes in the app. (Like SearchContextPicker on the home page)

## Tradeoffs 
- Since Combobox UI works only with unique strings we can't have dashboard with the same names in the options list. Because of this this PR enforces unique dashboard names by adding (1), (2), ... (N) suffix (example: Insights dashboard, Insights dashboard become Insights dashboard (1) and Insights dashboard (2)) 
- If you have selected dashboard and then you open the dashboard select popover you will see that search input has its name as a value. This is done for a better keyboard navigation (TL;DR it's reach-ui problem). This is the only way to enforce that keyboard navigation starts with selected element and not from the very beginning of the options list. More details in the reach-ui issues about it https://github.com/reach/reach-ui/issues/979

## Test plan
- Make sure that you can pick any dashboard that you can see in the dashboard select UI
- Check that all dashboard options are accessible from keyboard
- Check that dashboard select popover is placed in the right position in different screen sizes.
- Check that long dashboard names are handled properly (truncated) in select trigger button and in combobox options 



## App preview:

- [Web](https://sg-web-vk-improve-dashboard-select.onrender.com) Direct link: https://sg-web-vk-improve-dashboard-select.onrender.com/search
- [Storybook](https://5f0f381c0e50750022dc6bf7-mvttuyisdp.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
